repair_shards.c

```diff
@@ -115,14 +141,54 @@ master_copy_shard_placement(PG_FUNCTION_ARGS)
 
 
 /*
- * worker_copy_shard_placement is unimplemented but will connect to a source
- * shard placement in order to copy all rows from a shard table to an empty
- * shard table locally.
+ * worker_copy_shard_placement implements a internal UDF to copy a table's data from
+ * a healthy placement into a receiving table on an unhealthy placement. This
+ * function returns a boolean reflecting success or failure.
  */
 Datum
-worker_copy_shard_placement(PG_FUNCTION_ARGS __attribute__((unused)))
+worker_copy_shard_placement(PG_FUNCTION_ARGS)
 {
-	ereport(ERROR, (errmsg("worker_copy_shard_placement not implemented")));
+	text *shardRelationNameText = PG_GETARG_TEXT_P(0);
+	text *nodeNameText = PG_GETARG_TEXT_P(1);
+	int32 nodePort = PG_GETARG_INT32(2);
+	char *shardRelationName = text_to_cstring(shardRelationNameText);
+	char *nodeName = text_to_cstring(nodeNameText);
+	bool fetchSuccessful = false;
+
+	Oid shardRelationId = ResolveRelationId(shardRelationNameText);
+	Relation shardTable = heap_open(shardRelationId, RowExclusiveLock);
+	TupleDesc tupleDescriptor = RelationGetDescr(shardTable);
+	Tuplestorestate *tupleStore = tuplestore_begin_heap(false, false, work_mem);
+
+	StringInfo selectAllQuery = NULL;
+	ShardPlacement *placement = NULL;
+	Task *task = NULL;
+
+	selectAllQuery = makeStringInfo();
+	appendStringInfo(selectAllQuery, SELECT_ALL_QUERY,
+	                 quote_identifier(shardRelationName));
+
+	placement = (ShardPlacement *) palloc0(sizeof(ShardPlacement));
+	placement->nodeName = nodeName;
+	placement->nodePort = nodePort;
+
+	task = (Task *) palloc0(sizeof(Task));
+	task->queryString = selectAllQuery;
+	task->taskPlacementList = list_make1(placement);
+
+	fetchSuccessful = ExecuteTaskAndStoreResults(task, tupleDescriptor, tupleStore);
+	if (!fetchSuccessful)
+	{
+		ereport(ERROR, (errmsg("could not store shard rows from healthy placement"),
+						errhint("Consult recent messages in the server logs for "
+								"details.")));
+	}
+
+	CopyDataFromTupleStoreToRelation(tupleStore, shardTable);
```

Comment:
I agree somewhat, but there's also the method `CopyDataFromFinalizedPlacement` in this same file, so `CopyDataFromTupleStoreToRelation` is more analogous to it. I'll think about what I want to do (unsure right now).